Refactor vecgeom implementation to use collections

sethrj opened the pull request:
This will make it easier to add dual CPU/GPU implementations of apps (e.g. the rasterizer) and library functions.

Currently the last commit (the refactor) causes the tests to fail with
```
cudaErrorIllegalAddress: an illegal memory access was encountered
```
so I'll keep digging into this, but until then I could use an extra set of eyes on what I might be doing wrong/non-optimally.

mrguilima commented on src/geometry/detail/VGNavCollection.hh:
```diff
@@ -0,0 +1,182 @@
+/*!
+ * Manage a pool of device-side geometry states.
+ *
+ * Construction of the navstatepool has to be in a host compliation unit due to
```
typos: _compilation_ and `NavStatePool`.
(I could not figure out how to make the changes as a suggestion), as I've seen done by @sethrj)

mrguilima reviewed (approved):
Great to see the geometry using the same collections as physics -- I just need to get used to it :-)

sethrj commented:
@mrguilima Thanks for the feedback, I've just cleaned up the documentation for the VG nav collection. I was hoping specifically that you or @pcanal might be able to point out any less hacky ways for the VecGeom forward declarations, and even better if you guys could spot the flaw in 3b342d4bff7a627c8f35184c66477403ebde265f that's causing the vecgeom unit tests to crash (even though the rasterizer seems to be fine). I've added numerous debugging statements and checks, and compared with the prior "working" code, but I can't find the flaw in what's going on.

sethrj commented on src/geometry/GeoInterface.hh:
```diff
@@ -49,7 +68,7 @@ struct GeoParamsPointers
 /*!
  * Data required to initialize a geometry state.
```
Because a particle track has not just a geometry state but a physics state, particle state, etc. right? The geometry state is just a subset of a single track's state,  and the initializer there is the data required to initialize that single track's geometry state. Am I misunderstanding your confusion?

mrguilima replied:
Yes, it makes a lot more sense explained this way.  Thanks!

sethrj commented:
Update from @pcanal:

- The main index table seems okay but child info in new volume seems odd
- The index array is actually a convolution of indices, translation vector and rotation matrix (so odd looking indices are actually part of the double for a matrix) 
- vgg_test_kernel crashes (inside VecGeom) with a case where it is run concurrently only for tid==5 and tid==6  (but emulating running those 2 serially leads to a lack of crash)

pcanal commented:
See https://github.com/sethrj/celeritas/pull/1

sethrj commented:
Argggg thanks @pcanal , so sorry to have sent you on that wild goose chase.

There are a few software quality changes to VecGeom that could have prevented this mistake of mine and can prevent other mistakes in the future.
1. The vecgeom navstatepool API requires using extremely dangerous casts, doing raw pointer manipulation, etc. I think NavState needs to have an API for passing in a pool pointer and thread ID, and maybe even an overload that works with CPU as well. Because...
2. If the same code were being compiled on CPU and GPU, `-Wextra -Werror` would have caught the unused variable `result`.
3. If VecGeom fixed the `unknown stack size`  warnings we could run CUDA with `-Werror` and extra flags as well.

Anyway, thanks again for the find, sorry it took so long, and let's try to make improvements to the VecGeom API going forward to make it safer for GPU use. Speaking of which, did you have any other comments on this PR review? I have to do some similarly ugly `#ifdef`ing to force some code to compile on both NVCC and GCC and would appreciate if you could take another look and see if there's anything I can do safer and better.

pcanal commented on src/geometry/GeoInterface.hh:
```diff
@@ -32,14 +38,31 @@ namespace celeritas
+    using PlacedVolumeT = typename detail::VGTraits<M>::PlacedVolume;
+
+    const PlacedVolumeT* world_volume = nullptr;
+    int                  max_depth    = 0;
```
Those are "tied together", i.e. you probably don't want to modify one without the other.  Are the GeomParamsData "reused" or should those 2 members be `const`?

sethrj replied:
Everywhere, `Params` data are always set exactly once during the constructor, then the "reference" (which actually copy these non-Collection values) are copied on host, and the data's copied to device. We can't truly make the data const without adding extra copy constructors and going through a whole lot of other rigamarole -- but in practice (since the CollectionMirror is assigned only once) they're effectively const.

pcanal commented on src/geometry/detail/VGNavCollection.cc:
```diff
@@ -0,0 +1,97 @@
+                                                               size_type size)
+{
+    CELER_EXPECT(max_depth > 0);
+    CELER_EXPECT(size == 1);
```
This is surprising ... Is it documented why this particular collection can only be of size 1?

sethrj replied:
I can add to the docs -- but basically it looks like the VecGeom navstatecollection always allocates GPU memory when compiling with CUDA, so we can't use it for cuda-disabled tests (and prob don't want to anyway for efficiency). Since we're not currently running more than one simultaneous track at a time on CPU (no good reason to since we don't have threads) the limitation to 1 should be OK for now right?

pcanal replied:
Unless we are certain that we won't extent it, it would make more sense to prepare it to be `size>1` ready while the implementation is fresh in our memory ... humm actually nevermind, it looks like the implementation is small enough that it does not matter.

pcanal commented on src/geometry/detail/VGNavCollection.cc:
```diff
@@ -0,0 +1,97 @@
+    CELER_EXPECT(sz > 0);
+    CELER_EXPECT(celeritas::device());
+
+    pool.reset(new vecgeom::cxx::NavStatePool(sz, max_depth));
```
This is likely related to my previous question.  Why does the host contain a `NavState` and the device a `NavStatePool`?

sethrj replied:
I think I answered your question above -- navstatepool isn't very CPU-friendly.

pcanal commented on src/geometry/detail/VGNavCollection.hh:
```diff
@@ -0,0 +1,204 @@
+ * Get the navigation state at the given thread.
+ */
+CELER_FUNCTION auto VGNavCollection<Ownership::reference, MemSpace::device>::at(
+    int max_depth, ThreadId thread) const -> NavState&
```
Having the max_depth being 'customizable' is a source of potential problem.  The depth used here *must* be the same as the depth used during allocation.

sethrj replied:
Yeah, true -- originally I had `max_depth` as part of the state too but it's redundant with the params. It wouldn't be "costly" to duplicate that for the VG nav collection as member data (two more ints in const memory, no big deal) but it's definitely less elegant. I can add it as state data or just add a comment about making sure it's consistent -- your call.

pcanal replied:
it just that it is a potential source of mistake since the arguement "max_depth" *must* be the one expected or the result is wrong (I.e. it is a property of the collection not a parameter of the at call).

sethrj replied:
You're right Philippe -- to prevent mistakes I've added `max_depth` to the state data. Partly out of laziness (changing the interface) and partly out of potential optimization (max_depth being loaded once from 'params' rather than from the state in two separate places) I'm still passing in max_depth which is assumed to come from the Params class, but a new assertion will make sure they're consistent.

pcanal commented on src/geometry/detail/VGNavCollection.hh:
```diff
@@ -0,0 +1,204 @@
+    CELER_EXPECT(thread < size);
+#ifdef __NVCC__
+    // This code only compiles when run through CUDA so it must be escaped.
+    char* result = reinterpret_cast<char*>(this->ptr);
+    result += NavState::SizeOfInstanceAlignAware(max_depth) * thread.get();
+    return *reinterpret_cast<NavState*>(result);
```
As a side note this is the implementation of `NavigationPool::operator[]`.

At the moment it is not `practical` to use it as `vecgeom::cuda::NavStatePool` is missing a constructor (and destructor) to build one from already allocated memory.  

Side side note, if this class had the data members:
```
  int fCapacity; // the number of states in the pool
  int fDepth;    // depth of the navigation objects to cover
  char *fBuffer; // the memory buffer in which we place states
```
the following C++ (fragile) code would work:
```
return reinterpret_cast<vecgeom::NavStatePool>(this)->operator[thread.get()];
```

> // This code only compiles when run through CUDA so it must be escaped.

I am not sure why that is the case, the same code is in host-device function in NavStatePool ...

sethrj replied:
> ```c++
> return reinterpret_cast<vecgeom::NavStatePool>(this)->operator[thread.get()];
> ```
😱

pcanal commented on src/geometry/detail/VGNavCollection.hh:
```diff
@@ -0,0 +1,204 @@
+} // namespace vecgeom
+namespace vecgeom
+{
+#ifdef __NVCC__
+namespace cxx
+#else
+namespace cuda
+#endif
+{
+class NavigationState;
+}
```
This seems to stems from omissions in the VecGeom header files.  I will take a look.

pcanal commented:
> I have to do some similarly ugly `#ifdef`ing to force some code to compile on both NVCC and GCC and would appreciate if you could take another look and see if there's anything I can do safer and better

@sethrj Is there more than the one in VGNavCollection.hh  that you are worried about?

sethrj commented:
> @sethrj Is there more than the one in VGNavCollection.hh that you are worried about?

No, mostly the forward-declarations at the top for the raw pointers to `NavigationState`. They're technically wrong (because `NavigationState` is a compile time switch--dependent typedef) but I couldn't find a way around it. The `NavStatePool` I recall also having issues  -- it's not declared when including the file from NVCC.

These lack of forward declarations aren't surprising because of the way the `VECGEOM_NAMESPACE` macro stuff works, but I was hoping that the headers for the public-facing classes in VecGeom could correctly declare the non-`NAMESPACE` versions at the same time (e.g. declare non-inline `vecgeom::cuda` when compilling without `__NVCC__`, and declare non-inline `vecgeom::cxx` when compiling with NVCC).
